components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java

```diff
@@ -0,0 +1,165 @@
+
+        return new AuthServiceResponseWrapper(response);
+    }
+
+    private AuthServiceResponse processCommonAuthResponse(AuthServiceRequestWrapper request,
+                                                          AuthServiceResponseWrapper response)
+            throws AuthServiceException {
+
+        AuthServiceResponse authServiceResponse = new AuthServiceResponse();
+        if (isAuthFlowCompleted(request)) {
+            authServiceResponse.setFlowId(getFlowCompletionFlowId(request, response));
+            authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.SUCCESS_COMPLETED);
+            return authServiceResponse;
+        }
+        authServiceResponse.setFlowId(response.getFlowId());
+        authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.INCOMPLETE);
+        setAuthResponseData(request, response, authServiceResponse);
+
+        return authServiceResponse;
+    }
+
+    private void setAuthResponseData(AuthServiceRequestWrapper request, AuthServiceResponseWrapper response,
+                                     AuthServiceResponse authServiceResponse) throws AuthServiceException {
+
+        AuthServiceResponseData responseData = new AuthServiceResponseData();
+        boolean isMultiOptionsResponse = request.isMultiOptionsResponse();
+
+        List<AuthenticatorData> authenticatorDataList;
+        if (isMultiOptionsResponse) {
+            responseData.setAuthenticatorSelectionRequired(true);
+            authenticatorDataList = getAuthenticatorBasicData(response.getAuthenticators());
+        } else {
+            authenticatorDataList = request.getAuthInitiationData();
+        }
+        responseData.setAuthenticatorOptions(authenticatorDataList);
+        authServiceResponse.setData(responseData);
+    }
+
+    private List<AuthenticatorData> getAuthenticatorBasicData(String authenticatorList) {
+
+        List<AuthenticatorData> authenticatorDataList = new ArrayList<>();
+        String[] authenticatorAndIdpsArr = StringUtils.split(authenticatorList,
+                AuthServiceConstants.AUTHENTICATOR_SEPARATOR);
+        for (String authenticatorAndIdps : authenticatorAndIdpsArr) {
+            String[] authenticatorIdpSeperatedArr = StringUtils.split(authenticatorAndIdps,
+                    AuthServiceConstants.AUTHENTICATOR_IDP_SEPARATOR);
+            String name = authenticatorIdpSeperatedArr[0];
+            ApplicationAuthenticator authenticator = FrameworkUtils.getAppAuthenticatorByName(name);
+            if (authenticator == null) {
+                LOG.debug("Authenticator not found for name: " + name);
+                continue;
+            }
+
+            if (!authenticator.isAPIBasedAuthenticationSupported()) {
+                LOG.debug("Authenticator: " + name + " does not support API based authentication.");
+                continue;
+            }
+            for (int i = 1; i < authenticatorIdpSeperatedArr.length; i++) {
```

Comment:
are we skipping the 0th one in the list? better to add a small comment mentioning what is being done here